Added support for tracing via SNS message attributes

janderson007 opened the pull request:
Adds a request handler that can be added to a AWS SNS client to add tracing to the message attributes when publishing a message to a topic.

cc @llinder

devinsba commented:
FYI, we are tracking a similar feature for inclusion in `zipkin-aws`. I left the following comment with what I learned from looking over the X-Ray library: https://github.com/openzipkin/zipkin-aws/issues/53#issuecomment-346450852

Also, I know personally I'd be able to give time to get this merged into that repo if you wanted to try to make a similar PR there

janderson007 commented:
As far as I can tell, XRay is just generically tracing the client request and the service response; for SNS/SQS, it doesn't trace the actual message. The consumer that receives the message will not be able to continue the trace, and ultimately, that is what we wanted.

For this SNS tracing to work, it's relying on
1. a consumer that will pull the trace id from the `MessageAttributes` in order to continue the trace.  (We are subscribing via SQS, so this is handled via the instrumentation of SQS that @llinder put together)
2. `RawMessageDelivery` needs to be enabled for the SNS subscription, so that the `MessageAttributes` get propagated to the subscriber.  (The application is responsible for this)

So our use case is a little bit specific, and this PR by itself might not be all that useful without the SQS tracing as well.    I guess it depends if you want to be able to just trace through the API calls like XRay does, or if you want a deeper level of tracing that gets propagated through the actual messages.

devinsba commented:
Ah, I missed the detail that this was metadata attachment. Thanks for the clarification! I still think these are valuable instrumentations if you all ever decide to upstream them into the "official" repos

llinder commented:
Thanks for the feedback @devinsba!  Having general Zipkin instrumentation for the AWS clients would be awesome and it looks like you have some good thoughts on the implementation already.  As @janderson007 described, our use case is more about context propagation and not just API tracing but I think both have their place.

I'm fairly overwhelmed with work through mid January but will give this some thought and we can collaborate more via your document you started or on openzipkin Gitter channel.

llinder commented on brave-sns/src/main/java/smartthings/brave/sns/PublishRequestTracingHandler.java:
```diff
@@ -0,0 +1,92 @@
+ * AWS RequestHandler2 that adds tracing around the afterError and afterResponse events
+ * when calling the AWS SNS publish API.
+ */
+public class PublishRequestTracingHandler extends RequestHandler2 implements SNSRequestTracingHandler {
```
I haven't looked at the RequestHandler2 class much nor the X-Ray integrations with the AWS SDK.  From what I gather though we might want this to be a place where we hook in more Zipkin tracing details right?  If thats the case maybe this should be named `TracingRequestHandler` with the assumption more than just `PublishRequest` handling will be added here as well.

llinder replied:
Should be `final`

llinder commented on brave-sns/pom.xml:
```diff
@@ -0,0 +1,74 @@
+        </dependency>
+        <dependency>
+            <groupId>com.amazonaws</groupId>
+            <artifactId>aws-xray-recorder-sdk-aws-sdk</artifactId>
```
Its not super clear so I will ask :)  why they dependency on `aws-xray-recorder-sdk-aws-sdk`?

janderson007 replied:
Oops, shouldn't have been committed.  I removed it

janderson007 commented:
I added a README.md and implemented the suggestions. 

 I also moved the injection of the trace attributes from `beforeRequest` and into `beforeMarshalling`.    The request is actually marshalled  _before_ the `beforeRequest`, (I'd assumed the opposite)  so any changes to the original request in `beforeRequest` do **not** get marshalled.  

(The XRay library uses `beforeRequest`, but is directly setting HTTP Request headers rather than modifying the original request with message attributes)
